component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -758,28 +758,23 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
             }
             throw new InvalidCredentialsException("Retrying to resend the OTP");
         }
-
+        boolean succeededAttempt = false;
         if (userToken.equals(contextToken)) {
             processValidUserToken(context, authenticatedUser);
+            succeededAttempt = true;
         } else if (isLocalUser && "true".equals(SMSOTPUtils.getBackupCode(context))) {
-            checkWithBackUpCodes(context, userToken, authenticatedUser);
+            succeededAttempt = checkWithBackUpCodes(context, userToken, authenticatedUser);
         } else {
+            context.setProperty(SMSOTPConstants.CODE_MISMATCH, true);
+        }
+        if (!succeededAttempt) {
             handleSmsOtpVerificationFail(context);
-            handleCodeMismatch(context);
+            throw new AuthenticationFailedException("Code mismatch.");
```

Comment:
Don't we have to set `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` here?

Reply:
If the user token doesn't match with the code received to the mobile or doesn't match with any backup code `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` is set.
Addition to those cases, the `succeededAttempt` is failed for the following reasons,

- Code is expired

- The user hasn't any backup code when the user token is not matched with the code sent to the mobile.

So, We don't need to set `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)` here. 
I think we can change the error message. eg: `SMS OTP verification failure.` WDYT?

Reply:
AFAIU if the code has expired the program will not reach this point right? Therefore this line will only reach for a code mismatch right?
Seems like other methods handle `context.setProperty(SMSOTPConstants.CODE_MISMATCH, true)`. Therefore, we do not need to add it here.
Therefore, shall we change the message something like `Invalid code. Verification failed`? WDYT?

Reply:
Yes for SMS OTP, if the code expired it doesn't reach here, but in email OTP it happens.
+1 for `Invalid code. Verification failed`

Reply:
Fixed in a8b29a5